pkg/util/steps/runner.go

```diff
@@ -21,19 +21,29 @@ func FriendlyName(f interface{}) string {
 type Step interface {
 	run(ctx context.Context, log *logrus.Entry) error
 	String() string
+	MetricsTopic() string
 }
 
 // Run executes the provided steps in order until one fails or all steps
 // are completed. Errors from failed steps are returned directly.
-func Run(ctx context.Context, log *logrus.Entry, pollInterval time.Duration, steps []Step) error {
+// time cost for each step run will be recorded for metrics usage
+func Run(ctx context.Context, log *logrus.Entry, pollInterval time.Duration, steps []Step, metricsDryrun bool) (map[string]int64, error) {
+	stepTimeRun := make(map[string]int64)
 	for _, step := range steps {
 		log.Infof("running step %s", step)
+		startTime := time.Now()
 		err := step.run(ctx, log)
 
 		if err != nil {
 			log.Errorf("step %s encountered error: %s", step, err.Error())
-			return err
+			return nil, err
 		}
+
+		endTime := time.Now()
+		if metricsDryrun {
```

Comment:
IMO we should not modify production code in such a way to only be used for testing. The only place I see `true` sent for `metricsDryrun` is in tests. While this fixes the use of sleep from before it now adds complexity throughout the code base as we need to pass `false` in many places where we want to run steps.

Reply:
I'm thinking dependency injection is probably the best answer... create an interface that `time` can match (and is used in prod), but for the test, you make an implementation that returns reasonable values for `Now` and other functions.